src/lxc/lxccontainer.c

```diff
@@ -4659,6 +4680,158 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
 
 WRAP_API_2(bool, lxcapi_restore, char *, bool)
 
+static int do_lxcapi_mount(struct lxc_container *c,
+			const char *source, const char *target,
+			const char *filesystemtype, unsigned long mountflags,
+			const void *data, struct lxc_mount *mnt) {
+	int fd;
+	char *template, *suff, *path;
+	pid_t pid, init_pid;
+	size_t len;
+	struct stat sb;
+	int ret = -1;
+
+	if (!c || !c->lxc_conf) {
+		ERROR("Container or configuration is NULL");
+		goto out_err;
+	}
+
+	len = strlen(c->lxc_conf->lxc_shmount.path_host) + sizeof("/lxcmountXXXXXX");
+	template = malloc(len);
```

Comment:
Hm, no what I meant is:

```
len = strlen(c->lxc_conf->lxc_shmount.path_host) + sizeof("/lxcmountXXXXXX") - 1;
template = malloc(len + 1);
```
This way the right-hand side of the conjunct in the error check for `snprintf()` will make sense:
```
ret = snprintf(template, len, "%s/lxcmountXXXXXX", c->lxc_conf->lxc_shmount.path_host);
if (ret < 0 || (size_t)ret >= len)
```

Reply:
Ah, I finally see where I was mistaken. I saw the part of the `man snprintf` about writing `at most _size_ bytes (including the terminating null byte ('\0'))`, but I missed the `return the number of characters printed (excluding the null byte <...>)`.

Reply:
Hmm. Wait. 
In your code snippet, the length of the meaningful part of the string is `len`, and the length with the terminating null-byte is `len + 1`. `len + 1` bytes are `malloc`ed. Further, `snprintf` would write `<= len` bytes, and the null-byte would be assumed among these `len` (which would not be true).
The check, AFAIU, would make sense anyway as long as the parameter matches the one we passed to the `snprintf` function as __size__ .

Reply:
`len` is also used to tell `snprintf()` how many bytes it is supposed to write at maximum:

    snprintf(..., len, ...);

If `len` does include the count for the `\0` byte you get truncated results.

Reply:
The generic patter in the codebase is to calculate the length without the terminating zero byte since this is in line with what `strlen()` does. The terminating zero byte is only accounted for when you allocate a buffer.

Reply:
Please switch to the `malloc(len + 1)` style everywhere so that the length of a string means exactly what `strlen()` makes it mean: all bytes **excluding** the `\0` byte. :)